twim/calibration/R3BTwimMapped2Cal.cxx

```diff
@@ -191,97 +236,86 @@ void R3BTwimMapped2Cal::Exec(Option_t*)
         return;
 
     auto** mappedData = new R3BTwimMappedData*[nHits];
-    Int_t secId = 0;
-    Int_t anodeId = 0;
-    Double_t pedestal = 0.;
-    Double_t slope = 1.;
 
-    for (Int_t s = 0; s < fNumSec; s++)
-        for (Int_t i = 0; i < (fNumAnodes + fNumAnodesRef + fNumAnodesTrig); i++)
+    for (Int_t sec = 0; sec < fNumSec; sec++)
+    {
+        fTwimCal[sec]->Init();
+    }
+    for (Int_t i = 0; i < nHits; i++)
+    {
+        mappedData[i] = dynamic_cast<R3BTwimMappedData*>(fTwimMappedDataCA->At(i));
+        auto secId = mappedData[i]->GetSecID() - 1;
+        auto anodeId = mappedData[i]->GetAnodeID() - 1;
+        if (anodeId < fNumAnodes && fCal_Par->GetInUse(secId + 1, anodeId + 1) == 1)
         {
-            mulanode[s][i] = 0;
-            for (Int_t j = 0; j < fMaxMult; j++)
+            auto pedestal = CalEParams[secId]->GetAt(fNumEParams * anodeId);
+            auto slope = CalEParams[secId]->GetAt(fNumEParams * anodeId + 1);
+            fTwimCal[secId]->GetAnode(anodeId)->SetVal(pedestal + slope * mappedData[i]->GetEnergy(),
+                                                       mappedData[i]->GetTime());
```

Comment:
If I understand correctly, the fTwimCal is first zero initialised with a maximal amount of elements. And then fill the valuable values according to section ids and anode ids. 

If so, I would suggest to use std::map in your data structure:
```cpp
class CalSection {
    public:
    //...
    using AnodeMap = std::map<size_t, CalAnode>;
    private:
        AnodeMap fAnode;
        AnodeMap fTref;
};

std::map<size_t, CalSection> fTwimCal;
```

Latter when you fill the value:
```cpp
        auto secId = mappedData[i]->GetSecID() - 1;
        auto anodeId = mappedData[i]->GetAnodeID() - 1;
        
        auto anodes = AnodeMap{};
        auto anode = CalAnode{};
        anode.SetE();
        //...
        anodes.insert(anodeID, anode);

        //...
        fTwimCal.insert(sectionId, anodes);
```

Reply:
Thanks for the suggestion. std::map sounds interesting solution, but I need to handle multiple hits for the respective channel, thus the energy and timing information should be stored in array or vector format. Do I understand correctly that the example you showed can overwrite the numbers if there is more than one hit in a channel?

Reply:
Ok, in this case you need to have a check whether the AnodeMap has been inserted before. If no, insert it. If yes, just get it:

```cpp
if(fTwimCal.find(sectionId) != fTwimCal.end())
{
    fTwimCal.insert({sectionID, AnodeMap{}});
}
auto anodes = fTwimCal.at(sectionID);
if(anodes.find(anodeID) != anodes.end())
{
     anodes.insert({anodeID, CalAnode{}});
}
anodes.at(anodeID).GetHits().push_back(Hit);
```
This way is very memory efficient but not very good at performance (anyway, there is always a tradeoff between the performance and memory efficiency).